### PR TITLE
Caching of file-set hashes by local path and mtimes

### PR DESCRIPTION
**Pull request by tclose**
## Types of changes

- New feature (non-breaking change which adds functionality)

## Summary

Will address #683 by

- [x] `bytes_repr()` overloads for FileSets to yield a "time-stamp" object consisting of a key (file path) and modification time as the first item in the generator.
- [x] `hash_single()` checks for these key/mtime pairs in a global "local hashes cache" dict
    - returns cached hash, if present
    - otherwise it proceeds through the remaining byte chunks, and calculates the hash
- [x] calculated hashes are saved into/loaded from the cache directory using a hash of the key/mtime 

## Checklist

- [x] I have added tests to cover my changes (if necessary)
- [x] I have updated documentation (if necessary)

## Notes
I'm pretty happy with how this turned out with the exception of a few of wrinkles. Any suggestions would be most welcome
- There isn't a clean way to specify the location of the persistent hash cache in the top-level code or put it in the `cache_location` path (my original plan) given that `checksum` and `hash` are object properties instead of methods
    - I have therefore just dumped it in `~/.pydra-hash-cache`
    - not ideal to have to (effectively) hard-code this and having it user dependent
    - Having it in the user directory does mean that if the same cache directory is accessed from different machines (e.g. shared network drive) then local paths won't clash (although chance of mtimes being the same would be vanishingly small)
- The keys of the caches themselves are hashed and stored as files in the persistent cache directory, and will therefore never be cleaned up.
    - This could be ok as they are very small files so if it grows as time goes on it probably won't have much impact
    - This is how pydra caches work in general
- The resolution of mtime differs depending on OS, with Ubuntu and Windows having quite low resolution, i.e. order of seconds
    - I have had to put a sleep call inside the hash test to ensure the mtimes are different
    - Could cause issues if you were updating the contents of files within a loop and spinning off different workflows (not sure whether this would ever happen in practice)
    - No easy way to disable the mtime caching behaviour if this does become a problem

### CODE REVIEW

**Comment by ghisvail** [comment]
I am not confident I understand enough of the previous and proposed caching methods to have an opinion and assess this PR.

AFAIK, the previous caching mechanism generated cache folders for each task (including the workflow) in a common cache location which was set to a temporary folder, unless overridden with the `cache_dir` argument.

Each task's cache folder was composed of its name and a hash value, the latter being computed from the task's input field values. The cache would be re-used, if the name and hash values did not change, and the same cache folder was specified as `cache_dir`. Otherwise, it would recompute everything in a new cache location somewhere in temp.

Could you confirm whether my summary is accurate and whether this PR is proposing to change the fundamentals of this mechanism?

**Comment by tclose** [comment]
> I am not confident I understand enough of the previous and proposed caching methods to have an opinion and assess this PR.

No worries, I just put you all down as reviewers, so you were notified. Don't feel like you need to contribute if the area isn't familiar

> AFAIK, the previous caching mechanism generated cache folders for each task (including the workflow) in a common cache location which was set to a temporary folder, unless overridden with the `cache_dir` argument.
> 
> Each task's cache folder was composed of its name and a hash value, the latter being computed from the task's input field values. The cache would be re-used, if the name and hash values did not change, and the same cache folder was specified as `cache_dir`. Otherwise, it would recompute everything in a new cache location somewhere in temp.
> 
> Could you confirm whether my summary is accurate and whether this PR is proposing to change the fundamentals of this mechanism?

Yes, your understanding is correct. Tis is how the hashing works, both currently and in this PR. This PR seeks to improve (or more accurately, restore) performance by caching the hashes of file/directory types themselves, so files/directories don't need to be rehashed (which can be an expensive operation) each time the checksum is accessed.

It does this by hashing the path and mtime of the file/directory (not to be confused with the hashing of the file/directory contents), and using it as a key to look up a "hash cache" (not to be confused with the execution cache) that contains previously computed file/directory hashes.

**Comment by effigies** [comment]
> * I have had to put a sleep call inside the hash test to ensure the mtimes are different

Better to mock the call than to sleep. Here's an example where I've done that with `time.time()`:
https://github.com/nipy/nibabel/blob/5f37398a2f8211c175b798eb42298b963c693ae0/nibabel/tests/test_openers.py#L457-L464

**Comment by tclose** [comment]
> > * I have had to put a sleep call inside the hash test to ensure the mtimes are different
> 
> Better to mock the call than to sleep. Here's an example where I've done that with `time.time()`: https://github.com/nipy/nibabel/blob/5f37398a2f8211c175b798eb42298b963c693ae0/nibabel/tests/test_openers.py#L457-L464

Interesting, I assumed that the value for the mtime was controlled by the file-system not Python.

File: pydra/utils/hash.py
Comment (tclose) [reply pydra/utils/hash.py:80]: ok, sounds good

File: pydra/utils/hash.py
Comment (ghisvail) [inline pydra/utils/hash.py:154]: I believe we should be careful where to place stuff in the home directory.

We could follow the old "wild west" convention of placing files in a single `.pydra` folder at the root of `$HOME`, but this is generally considered sloppy, and sooner or later someone will complain and ask whether we can support established standards for their platform of choice.

Luckily, there is a [package](https://pypi.org/project/platformdirs/) implementing this already. I guess we would be concerned by this [section](https://platformdirs.readthedocs.io/en/latest/api.html#cache-directory).

**Review by ghisvail — CHANGES_REQUESTED** [review]
Some suggestions and a question mark regarding the cache path in the home folder. Maybe also worth having a look to mtime mocking instead of sleep steps as suggested by Chris.

Looks solid otherwise 👍

**Comment by djarecka** [comment]
@tclose - you can try with new slurm testing workflow, should work now!

File: pydra/utils/hash.py
Comment (djarecka) [reply pydra/utils/hash.py:154]: - perhaps we could just use a specific environmental variable if specified by the user?
- Since this is meant to be used within specific run, should we have option to clean the files that are e.g. a week old?

**Review by djarecka — APPROVED** [review]
lgtm! just left comments regarding the location of the cache directory

File: pydra/utils/hash.py
Comment (tclose) [reply pydra/utils/hash.py:113]: I believe this is a strange limitation of attrs, that converters can't be classmethods. I presume it is in order to preserve performance

File: pydra/utils/hash.py
Comment (tclose) [reply pydra/utils/hash.py:154]: That package looks good @ghisvail, although I not sure this is worth adding another dependency for, what are everyone's thoughts?

In either case, an env var is a good idea @djarecka. Do we have a naming convention for them?

> Since this is meant to be used within specific run, should we have option to clean the files that are e.g. a week old?

I don't see this needing to be constrained to a specific run, but probably no harm to run a clean-up of week/month-old files after a workflow completes

File: pydra/utils/hash.py
Comment (tclose) [reply pydra/utils/hash.py:113]: Actually, I had this wrong. It is just a limitation of mypy. I have added it as a classmethod and just ignored the mypy warning

**Comment by tclose** [comment]
> > * I have had to put a sleep call inside the hash test to ensure the mtimes are different
> 
> Better to mock the call than to sleep. Here's an example where I've done that with `time.time()`: https://github.com/nipy/nibabel/blob/5f37398a2f8211c175b798eb42298b963c693ae0/nibabel/tests/test_openers.py#L457-L464

@effigies I'm not sure this works on Windows (it seems to work on Ubuntu)

File: pydra/utils/hash.py
Comment (ghisvail) [reply pydra/utils/hash.py]: > although I not sure this is worth adding another dependency for

Every piece of software which cache stuff to home end up writing (usually badly) some subset of what's covered in `platformdirs`. It's not worth adding if we are confident to implement it *right*. If not, the package is dependency free and [well tested](https://github.com/platformdirs/platformdirs/actions/workflows/check.yml), so the risks and additional bloat to Pydra would be minimal.

> we could just use a specific environmental variable if specified by the user

We could indeed (say `PYDRA_CACHE_DIR`), but the suitable fallback value is platform dependent. That's where the value of `platformdirs` comes in.

> I don't see this needing to be constrained to a specific run, but probably no harm to run a clean-up of week/month-old files after a workflow completes

I am trying to picture a use case for this: if I run workflow A and months later workflow B, should I expect Pydra to clean cached files up from workflow A (and any other which are old enough w.r.t. workflow B)?

Comment (tclose) [reply pydra/utils/hash.py]: > > although I not sure this is worth adding another dependency for
> 
> Every piece of software which cache stuff to home end up writing (usually badly) some subset of what's covered in `platformdirs`. It's not worth adding if we are confident to implement it _right_. If not, the package is dependency free and [well tested](https://github.com/platformdirs/platformdirs/actions/workflows/check.yml), so the risks and additional bloat to Pydra would be minimal.

Sounds ok to me, I looked into pulling out just the bits we need from that package, and for Mac+Linux it was trivial but Windows is pretty complicated so decided against it.

> > we could just use a specific environmental variable if specified by the user
> 
> We could indeed (say `PYDRA_CACHE_DIR`), but the suitable fallback value is platform dependent. That's where the value of `platformdirs` comes in.

👍

> 
> > I don't see this needing to be constrained to a specific run, but probably no harm to run a clean-up of week/month-old files after a workflow completes
> 
> I am trying to picture a use case for this: if I run workflow A and months later workflow B, should I expect Pydra to clean cached files up from workflow A (and any other which are old enough w.r.t. workflow B)?

Yes, if they are not touched by Workflow B and they are over a month old they get cleaned up.

Comment (tclose) [reply pydra/utils/hash.py]: It is still showing up as "1 change requested" from you, I think it might have been the method name change (which I implemented in the last round of changes). Is that right?

Comment (tclose) [reply pydra/utils/hash.py]: @ghisvail I realised that *platformdirs* is already dependency of *black* and *virtualenv* so have used it to locate the hash cache as you suggested.

**Comment by tclose** [comment]
@effigies, I think that I have addressed the outstanding issues with this PR so it just needs your review. However, I had to revert the mtime mocking as it doesn't appear to work for Windows (see https://github.com/nipype/pydra/actions/runs/6307685788/job/17124695852), unless you have some ideas on how to do it.

**Comment by djarecka** [comment]
@tclose - the recent commit to this PR comes from some rebasing?

**Comment by tclose** [comment]
> @tclose - the recent commit to this PR comes from some rebasing?

I rebased it on top of master after the environment PR was merged and fixed up a few things

**Comment by djarecka** [comment]
ok, trying to figure out if I should review it again, since I already accepted

File: pydra/utils/hash.py
Comment (effigies) [inline pydra/utils/hash.py:180]: `self.persistent` never gets used. How is this supposed to work?

**Review by effigies — COMMENTED** [review]
Sorry, I've completely lost the plot on this PR. I don't understand what's going on, but there are some things that look wrong to me.

File: pydra/utils/hash.py
Comment (tclose) [reply pydra/utils/hash.py:138]: I think I just copied it from somewhere. I don't really know the pros & cons of different methods. Which would you recommend?

File: pydra/utils/hash.py
Comment (tclose) [reply pydra/utils/hash.py:182]: It is used on L241 (https://github.com/tclose/pydra/blob/a031ea5dfc1571d2171e0df4157bb97114b9b1eb/pydra/utils/hash.py#L241)

File: pydra/utils/hash.py
Comment (djarecka) [reply pydra/utils/hash.py:139]: looks like we were using `blake2s` before, but indeed doesn't seem to be the best choice (probably copied from somewhere). We can just use `blake2 ` or try [blake3](https://github.com/BLAKE3-team/BLAKE3), but could be also done in a separate pr

**Comment by djarecka** [comment]
sorry, I'm sure we discussed it at some point, but I'm not sure about one important thing...  looks like if I move file around my filesystem I have no way of reusing the previous tasks now, is that right?

**Comment by djarecka** [comment]
sorry, I was wrong, the task has correct hash when the file is the same with a different path. 

Perhaps you can add this test: https://github.com/djarecka/pydra/blob/db782c20890797bd58eb1d52545b7104f7d41aa4/pydra/engine/tests/test_node_task.py#L1568

(I was planning to create a PR to you, but I must have merged to the branch more things to my branch)

**Comment by tclose** [comment]
> sorry, I was wrong, the task has correct hash when the file is the same with a different path.
> 
> Perhaps you can add this test: https://github.com/djarecka/pydra/blob/db782c20890797bd58eb1d52545b7104f7d41aa4/pydra/engine/tests/test_node_task.py#L1568
> 
> (I was planning to create a PR to you, but I must have merged to the branch more things to my branch)

Ok, nice addition

**Comment by djarecka** [comment]
I've just realized that we should also have a similar test when the persistent cache is used, but realized that it's not enough to set `PYDRA_HASH_CACHE`. Sorry, for asking for more work, but can we have some test when the persistent cache is used together with running task or workflow

**Comment by tclose** [comment]
@djarecka I have added a new test to ensure that the persistent cache gets hit during the running of tasks. Let me know if there is anything else you need me to do

File: pydra/engine/tests/test_node_task.py
Comment (djarecka) [inline pydra/engine/tests/test_node_task.py:1635]: @tclose - where are you setting the persistent cach path? i.e. `PYDRA_HASH_CACHE`

Comment (tclose) [reply pydra/engine/tests/test_node_task.py:1635]: At @ghisvail's suggestion, the hash cache is stored in a system-dependent user cache directory using `platformdirs.user_cache_dir` by default (e.g. `/Users/<username>/Library/Caches/pydra/<version-number>` on MacOS).

I have just tweaked the code so that it is now put in a `hashes` subdirectory of that user cache dir (accessible in the `pydra.utils.user_cache_dir` variable) just in case any other cache data needs to be stored at some point in the future.

Comment (djarecka) [reply pydra/engine/tests/test_node_task.py:1635]: oh, got it! Sorry I missed that. I've just realized that I made a typo when I was setting `PYDRA_HASH_CACHE` and that's why it was not saving the hashes there, and was confused where this is being saved..

**Comment by djarecka** [comment]
@tclose - thanks so much for the work!

**Pull request merged**
